### PR TITLE
Support for CRD-specific watch namespaces configuration

### PR DESCRIPTION
**Pull request by levan-m**
### What does this PR do?

CECO-1021

**For Reviewer: first commit is actual change; second commit is mostly refactor.**

controller-runtime 0.16 allows cache configuration on resource level. This change adds CRD-specific watch namespace env vars and sets them on corresponding CRDs or related resources. 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: vX.Y.Z
* Cluster Agent: vX.Y.Z

### Describe your test plan

Test using DatadogMonitors CRD, use following manifests to create monitors in two different namespaces.

* Namespace `monitor1`
  ```yaml
  apiVersion: datadoghq.com/v1alpha1
  kind: DatadogMonitor
  metadata:
    name: datadog-monitor-test1
    namespace: monitor1
  spec:
    query: "avg(last_10m):avg:system.disk.in_use{*} by {host} > 0.5"
    type: "metric alert"
    name: "Test monitor made from DatadogMonitor monitor1"
    message: "We are running out of disk space!"
    tags:
      - "test:datadog"
  ```

* Namespace `monitor2`
  ```yaml
  apiVersion: datadoghq.com/v1alpha1
  kind: DatadogMonitor
  metadata:
    name: datadog-monitor-test2
    namespace: monitor2
  spec:
    query: "avg(last_10m):avg:system.disk.in_use{*} by {host} > 0.5"
    type: "metric alert"
    name: "Test monitor made from DatadogMonitor monitor2"
    message: "We are running out of disk space!"
    tags:
      - "test:datadog"
  ```

#### Operator doesn't watch `namespace1` or `namespace2`.
1. Run operator with `datadogMonitorEnabled` set to true (with API, APP key set), running in `datadog` namespace.
2. Set default watch namespaces env variable to current namespace
```yaml
        - name: WATCH_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
```

Operator shouldn't reconcile the monitors from other namespaces (nothing logged or created).

#### Add `monitor1` namespace
1. In above setup add following env variable 
  ```yaml
        - name: DD_MONITOR_WATCH_NAMESPACE
          value: "monitor1"
  ``` 
Observe monitor from `monitor1` namespace getting reconciled - check logs and datadog app to confirm it's created.

2. Add `monitor2` to above env var 

Observer monitor in `monitor2` reconciled as well. 

Write there any instructions and details you may have to test your PR.

### Checklist

- [ ] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [ ] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: docs/configuration.v1alpha1.md [hunk 1]
```diff
@@ -148,13 +150,13 @@ spec:
...
 | agent.config.securityContext.seccompProfile.type | type indicates which kind of seccomp profile will be applied. Valid options are:  Localhost - a profile defined in a file on the node should be used. RuntimeDefault - the container runtime default profile should be used. Unconfined - no profile should be applied. |
-| agent.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID.  If unspecified, no groups will be added to any container. Note that this field cannot be set when spec.os.name is windows. |
+| agent.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the uid of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the uid of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```
Comment (brett0000FF) [inline docs/configuration.v1alpha1.md:155]:
```suggestion
| agent.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the UID of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the UID of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```

File: docs/configuration.v1alpha1.md [hunk 2]
```diff
@@ -327,13 +333,13 @@ spec:
...
-| clusterAgent.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID.  If unspecified, no groups will be added to any container. Note that this field cannot be set when spec.os.name is windows. |
+| clusterAgent.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the uid of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the uid of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```
Comment (brett0000FF) [inline docs/configuration.v1alpha1.md:338]:
```suggestion
| clusterAgent.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the UID of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the UID of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```

File: docs/configuration.v1alpha1.md [hunk 3]
```diff
@@ -414,13 +421,13 @@ spec:
...
-| clusterChecksRunner.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID.  If unspecified, no groups will be added to any container. Note that this field cannot be set when spec.os.name is windows. |
+| clusterChecksRunner.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the uid of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the uid of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```
Comment (brett0000FF) [inline docs/configuration.v1alpha1.md:426]:
```suggestion
| clusterChecksRunner.config.securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the UID of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the UID of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```

File: docs/configuration.v2alpha1.md [hunk 4]
```diff
@@ -355,13 +357,13 @@ In the table, `spec.override.nodeAgent.image.name` and `spec.override.nodeAgent.
...
 | [key].securityContext.seccompProfile.type | type indicates which kind of seccomp profile will be applied. Valid options are:  Localhost - a profile defined in a file on the node should be used. RuntimeDefault - the container runtime default profile should be used. Unconfined - no profile should be applied. |
-| [key].securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID.  If unspecified, no groups will be added to any container. Note that this field cannot be set when spec.os.name is windows. |
+| [key].securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the uid of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the uid of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```
Comment (brett0000FF) [inline docs/configuration.v2alpha1.md:362]:
```suggestion
| [key].securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the UID of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the UID of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```

**Review by brett0000FF — APPROVED** [review]
I know you're still drafting this, but I left a couple very minor suggestions in the meantime. Please let Docs know if add more content and need another look!

File: main.go [hunk 5]
```diff
@@ -237,7 +231,7 @@ func run(opts *options) error {
 		LeaseDuration:              &opts.leaderElectionLeaseDuration,
 		RenewDeadline:              &renewDeadline,
 		RetryPeriod:                &retryPeriod,
-		Cache:                      cacheOptions(setupLog),
+		Cache:                      config.CacheOptions(setupLog, opts.datadogAgentEnabled, opts.datadogAgentProfileEnabled, opts.datadogMonitorEnabled, opts.datadogSLOEnabled),
```
Comment (celenechang) [inline main.go:234]:
```suggestion
		Cache:                      config.CacheOptions(setupLog, opts.datadogAgentEnabled, opts.datadogMonitorEnabled, opts.datadogSLOEnabled, opts.datadogAgentProfileEnabled),
```

File: docs/configuration.v2alpha1.md (hunk 4 above)
Comment (levan-m) [reply docs/configuration.v2alpha1.md]: These files weren't changed by this PR and all suggestions are for docstrings from the dependencies. We don't have any control over those.

File: pkg/config/config.go [hunk 6]
```diff
@@ -6,18 +6,38 @@
...
-	// which specifies the Namespace to watch.
+	// AgentWatchNamespaceEnvVar is a comma separated list of namespaces watched by Agent controller.
+	agentWatchNamespaceEnvVar = "AGENT_WATCH_NAMESPACE"
+	// SLOWatchNamespaceEnvVar is a comma separated list of namespaces SLO controller.
+	sloWatchNamespaceEnvVar = "SLO_WATCH_NAMESPACE"
+	// MonitorWatchNamespaceEnvVar is a comma separated list of namespaces Monitor controller.
+	monitorWatchNamespaceEnvVar = "MONITOR_WATCH_NAMESPACE"
+	// ProfilesWatchNamespaceEnvVar is a comma separated list of namespaces Agent Profile controller.
+	profileWatchNamespaceEnvVar = "PROFILE_WATCH_NAMESPACE"
+	// WatchNamespaceEnvVar is a comma separated list of namespaces watched by all controllers unless controller specific configuration is provided.
```
Comment (celenechang) [inline pkg/config/config.go:38]: what do you think about prefixing with `DD_` consistent with how we define other envvars?

```suggestion
	// AgentWatchNamespaceEnvVar is a comma-separated list of namespaces watched by the DatadogAgent controller.
	agentWatchNamespaceEnvVar = "DD_AGENT_WATCH_NAMESPACE"
	// SLOWatchNamespaceEnvVar is a comma-separated list of namespaces watched by the DatadogSLO controller.
	sloWatchNamespaceEnvVar = "DD_SLO_WATCH_NAMESPACE"
	// MonitorWatchNamespaceEnvVar is a comma-separated list of namespaces watched by the DatadogMonitor controller.
	monitorWatchNamespaceEnvVar = "DD_MONITOR_WATCH_NAMESPACE"
	// ProfilesWatchNamespaceEnvVar is a comma-separated list of namespaces watched by the DatadogAgentProfile controller.
	profileWatchNamespaceEnvVar = "DD_AGENT_PROFILE_WATCH_NAMESPACE"
	// WatchNamespaceEnvVar is a comma-separated list of namespaces watched by all controllers, unless a controller-specific configuration is provided.
```

File: pkg/config/config.go [hunk 7]
```diff
@@ -30,38 +50,125 @@ const (
...
+	if datadogSLOEnabled {
+		sloNamespaces := getWatchNamespacesFromEnv(logger, sloWatchNamespaceEnvVar)
+		logger.Info("DatadogSLO Enabled", "watching namespaces", maps.Keys(sloNamespaces))
+		byObject[&datadoghqv1alpha1.DatadogSLO{}] = cache.ByObject{
+			Namespaces: sloNamespaces,
+		}
 	}
 
-	return []string{ns}
+	if datadogAgentProfileEnabled {
```
Comment (celenechang) [inline pkg/config/config.go:82]: should we add these cache filters if introspection is enabled?

File: pkg/config/config.go [hunk 8]
```diff
@@ -30,38 +50,125 @@ const (
...
-		// https://github.com/kubernetes-sigs/controller-runtime/commit/3e35cab1ea29145d8699259b079633a2b8cfc116
-		nsConfigs := make(map[string]cache.Config)
-		for _, ns := range namespaces {
-			nsConfigs[ns] = cacheConfig
+func getWatchNamespacesFromEnv(logger logr.Logger, envVar string) map[string]cache.Config {
+	cacheConfig := cache.Config{}
+
+	nsEnvValue, found := os.LookupEnv(envVar)
+	if !found {
+		logger.Info(fmt.Sprintf("CRD specific cache namespaces env variable %s not set, will be using common config", envVar))
```
Comment (celenechang) [inline pkg/config/config.go:155]:
```suggestion
		logger.Info(fmt.Sprintf("CRD-specific namespaces environmental variable %s not set, will be using common config", envVar))
```

File: pkg/config/config.go [hunk 9]
```diff
@@ -30,38 +50,125 @@ const (
...
+		nsEnvValue, found = os.LookupEnv(watchNamespaceEnvVar)
+		if !found {
+			logger.Info("Common namespace env variable %s not set, will be watching all namespaces", watchNamespaceEnvVar)
```
Comment (celenechang) [inline pkg/config/config.go:158]:
```suggestion
			logger.Info("Common namespaces environmental variable %s not set, will be watching all namespaces", watchNamespaceEnvVar)
```

File: pkg/config/config.go [hunk 10]
```diff
@@ -30,38 +50,125 @@ const (
...
 	}
-	logger.Info("Manager will watch and manage resources in all namespaces")
-	return map[string]cache.Config{cache.AllNamespaces: cacheConfig}
+
+	var namespaces []string
+	if strings.Contains(nsEnvValue, ",") {
+		namespaces = strings.Split(nsEnvValue, ",")
+	} else {
+		namespaces = []string{nsEnvValue}
+	}
```
Comment (celenechang) [inline pkg/config/config.go:168]: could we simplify this to 
```suggestion
	namespaces := strings.Split(nsEnvValue, ",")
```
?

File: pkg/config/config.go (hunk 6 above)
Comment (levan-m) [reply pkg/config/config.go:38]: I just followed the pattern of existing `WATCH_NAMESPACE` env var. Using DD makes sense with the new ones.

File: pkg/config/config.go (hunk 10 above)
Comment (levan-m) [reply pkg/config/config.go:168]: Thanks!

File: pkg/config/config.go (hunk 7 above)
Comment (levan-m) [reply pkg/config/config.go]: Yes, thanks for pointing this out. Fixed in 9th commit.

File: pkg/config/config.go [hunk 11]
```diff
@@ -30,38 +50,139 @@ const (
...
-			nsConfigs[ns] = cacheConfig
+func getWatchNamespacesFromEnv(logger logr.Logger, envVar string) map[string]cache.Config {
+	cacheConfig := cache.Config{}
+
+	nsEnvValue, found := os.LookupEnv(envVar)
+	if !found {
+		logger.Info(fmt.Sprintf("CRD-specific namespaces environmental variable %s not set, will be using common config", envVar))
+		nsEnvValue, found = os.LookupEnv(watchNamespaceEnvVar)
+		if !found {
+			logger.Info("Common namespaces environmental variable %s not set, will be watching all namespaces", watchNamespaceEnvVar)
```
Comment (khewonc) [inline pkg/config/config.go:177]:
```suggestion
			logger.Info(fmt.Sprintf("Common namespaces environmental variable %s not set, will be watching all namespaces", watchNamespaceEnvVar))
```

File: pkg/config/config.go [hunk 12]
```diff
@@ -30,38 +50,142 @@ const (
...
+		logger.Info("DatadogAgentProfile Enabled", "watching namespace", maps.Keys(agentProfileNamespaces))
+		byObject[profileObj] = cache.ByObject{
+			Namespaces: agentProfileNamespaces,
+		}
+
+		// It is very important to reduce memory usage when profiles are used.
+		// For the profiles feature we need to list the agent pods, but we're only
+		// interested in the node name and the labels. This function removes all the
+		// rest of fields to reduce memory usage.
+		// Pods are watched in DatadogAgent namespace(s) since that's were Agent pods are running.
```
Comment (celenechang) [inline pkg/config/config.go:110]:
```suggestion
		// Pods are watched in DatadogAgent namespace(s) since that's where Agent pods are running.
```

**Review by khewonc — APPROVED** [review]

**Pull request merged**